saexec/execution.go

```diff
@@ -0,0 +1,193 @@
+		scratch.statedb.SetTxContext(tx.Hash(), ti)
+
+		receipt, err := core.ApplyTransaction(
+			e.chainConfig,
+			dummy.ChainContext(),
+			&header.Coinbase,
+			&gasPool,
+			scratch.statedb,
+			header,
+			tx,
+			(*uint64)(&blockGasConsumed),
+			vm.Config{},
+		)
+		if err != nil {
+			// This almost certainly means that the worst-case block inclusion
+			// has a bug.
+			logger.Error(
+				"Transaction execution errored (not reverted)",
+				zap.Int("tx_index", ti),
+				zap.Stringer("tx_hash", tx.Hash()),
+				zap.Error(err),
+			)
+			continue
+		}
+
+		perTxClock.Tick(gas.Gas(receipt.GasUsed))
+		b.SetInterimExecutionTime(perTxClock)
+		// TODO(arr4n) investigate calling the same method on pending blocks in
+		// the queue. It's only worth it if [blocks.LastToSettleAt] regularly
+		// returns false, meaning that execution is blocking consensus.
+
+		// The [types.Header] that we pass to [core.ApplyTransaction] is
+		// modified to reduce gas price from the worst-case value agreed by
+		// consensus. This changes the hash, which is what is copied to receipts
+		// and logs.
+		receipt.BlockHash = b.Hash()
+		for _, l := range receipt.Logs {
+			l.BlockHash = b.Hash()
+		}
+
+		// TODO(arr4n) add a receipt cache to the [executor] to allow API calls
+		// to access them before the end of the block.
+		receipts[ti] = receipt
+	}
+	endTime := time.Now()
+	hook.AfterBlock(e.hooks, scratch.statedb, b.EthBlock(), e.gasClock, blockGasConsumed, receipts)
+	if e.gasClock.Time.Compare(perTxClock) != 0 {
+		return fmt.Errorf("broken invariant: block-resolution clock @ %s does not match tx-resolution clock @ %s", e.gasClock.String(), perTxClock.String())
+	}
+
+	logger.Debug(
+		"Block execution complete",
+		zap.Uint64("gas_consumed", uint64(blockGasConsumed)),
+		zap.Time("gas_time", e.gasClock.AsTime()),
+		zap.Time("wall_time", endTime),
+	)
+
+	root, err := e.commitState(scratch, b.NumberU64())
+	if err != nil {
+		return err
+	}
+	// The strict ordering of the next 3 calls guarantees invariants that MUST
+	// NOT be broken:
+	//
+	// 1. [blocks.Block.MarkExecuted] guarantees disk then in-memory changes.
+	// 2. Internal indicator of last executed MUST follow in-memory change.
+	// 3. External indicator of last executed MUST follow internal indicator.
+	if err := b.MarkExecuted(e.db, e.gasClock.Clone(), endTime, header.BaseFee, receipts, root); err != nil {
+		return err
+	}
+	e.lastExecuted.Store(b)                           // (2)
+	e.sendPostExecutionEvents(b.EthBlock(), receipts) // (3)
+	return nil
+}
+
+func (e *Executor) commitState(scratch *executionScratchSpace, blockNum uint64) (common.Hash, error) {
+	root, err := scratch.statedb.Commit(blockNum, true)
+	if err != nil {
+		return common.Hash{}, fmt.Errorf("%T.Commit() at end of block %d: %w", scratch.statedb, blockNum, err)
+	}
+
+	db, err := state.New(root, e.stateCache, scratch.snaps)
+	if err != nil {
+		return common.Hash{}, err
+	}
+	scratch.statedb = db
+	return root, nil
+}
```

Comment:
TODO: The handling of executionScratchSpace seems weird to me. I want to look into how this is structured.